### DISCUSSION TITLE
Using the Scene Understanding Extension

### DISCUSSION THREAD

**Discussion by laultman** [StereoKit/StereoKit#820] [Q&A] [answered]
@maluoi I would like to take advantage of more of the Scene Understanding extension capabilities and Windows spatial perception built-in functionality. You have some mention of this in your docs.

`World.RaycastEnabled must be set to true first! SK.System.worldRaycastPresent must also be true. This does a ray intersection with whatever represents the environment at the moment! In this case, it’s a watertight collection of low resolution meshes calculated by the Scene Understanding extension, which is only provided by the Microsoft HoloLens runtime.`

Could you point me in the right direction? I've been cruising through your C++ looking for possible implementations. I'm looking for sample code and trying to understand the Scene Understanding extension as available in SK.

I am trying to find a previously "mapped" object (my car in my garage) and later when I have moved the car, relocate all the mapped points on the car. Ultimately this will be used in training applications where the "world" is not the same, but the focus of the training is the same. Simply put - I want the car, not the world.

**Comment by maluoi** [comment, marked as answer]
Sounds like you're looking more for something like [Azure Object Anchors](https://azure.microsoft.com/en-us/products/object-anchors)! Alternatively if your training is constrained, a QR code could be the most reliable.

Scene Understanding is more about identifying common features of a room, but so far doesn't get more complex than labels for maybe "floor, door, table". Right now SK only uses Scene Understanding for raw room meshes, though at this point there's a few more extensions around, so I might be able to get to a few of those labeling features.

**Comment by laultman** [comment]
@maluoi Sadly, there are no [mixed reality services available in Azure Government](https://azure.microsoft.com/en-us/explore/global-infrastructure/products-by-region/?products=all&regions=non-regional%2Cusgov-non-regional%2Cus-dod-central%2Cus-dod-east%2Cusgov-arizona%2Cusgov-texas%2Cusgov-virginia&rar=true), so Azure Object Anchors is out. I must create my own solution. For commercial customers, Azure Object Anchors is the ticket!
Microsoft has only [limited features in Azure GCC](https://learn.microsoft.com/en-us/dynamics365/mixed-reality/guides/gcc) for Dynamics Guides and none in Azure Government/DoD. There is just not enough utilization in Azure Gov or DoD of these services to justify, the cost of obtaining an ATO, and you must have a DoD sponsor with funding to even begin the process. Microsoft can't unilaterally get an ATO on any DoD service.

**Comment by LarryAultman** [comment]
@maluoi You mentioned your use of Scene Understanding. Can you tell me more about how you are implementing it? I'd like to dig into it more. Thanks!

**Reply by maluoi** [reply]
SK's implementation of the Microsoft scene understanding ext [lives over here](https://github.com/StereoKit/StereoKit/blob/master/StereoKitC/xr_backends/openxr_scene_understanding.cpp). Meshing only, no semantic surfaces or anything like that. At the time when I wrote it, there wasn't much support for this sort of thing on other devices, so I was keeping it pretty simple until more universal consensus on the API could arrive. At a top level, these features show up in the `World` class.

Right now, there's at least 3 different Scene Understanding-like extensions for different platforms, including [`XR_EXT_plane_detection`](https://registry.khronos.org/OpenXR/specs/1.0/html/xrspec.html#XR_EXT_plane_detection). So it may be a little easier now to settle on some common API that covers some of the semantic labeling functionality.

**Comment by LarryAultman** [comment]
Thanks, I'll look more at this.
